lib/src/signed_video_h26x_auth.c

```diff
@@ -115,6 +115,112 @@ decode_sei_data(signed_video_t *self, const uint8_t *payload, size_t payload_siz
   return status;
 }
 
+/**
+ * Computes the GOP hash from |nalu_hash_list| and compares it with the GOP hash received from the
+ * signature.
+ *
+ * @param self A pointer to the signed_video_t structure.
+ * @return true if the computed GOP hash matches the received GOP hash and the status is SV_OK;
+ * false otherwise.
+ */
+static bool
+verify_gop_hash(signed_video_t *self)
+{
+  gop_info_t *gop_info = self->gop_info;
+  const size_t hash_size = self->verify_data->hash_size;
+  svrc_t status = SV_UNKNOWN_FAILURE;
+
+  // Compute the GOP hash from |nalu_hash_list|.
+  status = compute_partial_gop_hash(
+      self, gop_info->nalu_hash_list, gop_info->nalu_list_idx, gop_info->computed_gop_hash);
+  // Compare the computed GOP hash with the received GOP hash.
+  bool hashes_match =
+      (memcmp(gop_info->computed_gop_hash, self->received_gop_hash, hash_size) == 0);
+
+  // Verify the GOP hash if both the hashes match and the status is SV_OK.
+  gop_info->successfully_verified_gop_hash = (hashes_match && (status == SV_OK));
+
+  return gop_info->successfully_verified_gop_hash;
+}
+
+/**
+ * Adds the hashes of NALUs used in computing the GOP hash and marks these NALUs as part of the GOP
+ * hash.
+ *
+ * This function iterates through the NALU list, identifies the NALUs that belong to the current
+ * GOP, and marks them as used in the GOP hash. It then copies the hashes of these NALUs to the
+ * `nalu_hash_list` for later verification of the GOP hash. Additionally, it handles specific cases
+ * such as moving linked hash values for I-frames and ensures that SEI-related NALUs are
+ * appropriately handled.
+ */
+void
+prepare_for_verify_link_and_gop_hashes(signed_video_t *self, h26x_nalu_list_item_t *sei)
+{
+  // Ensure the `self` pointer is valid
+  assert(self);
+
+  // Initialize pointers and variables
+  h26x_nalu_list_t *nalu_list = self->nalu_list;
+  uint8_t *hash_list = &self->gop_info->nalu_hash_list[0];
+  int *list_idx = &self->gop_info->nalu_list_idx;
+  assert(nalu_list);
+  const size_t hash_size = self->verify_data->hash_size;
+  h26x_nalu_list_item_t *item = NULL;
+  uint8_t *hash_to_add = NULL;
+  bool found_next_gop = false;
+  bool found_item_after_sei = false;
+  uint8_t linked_hashes[hash_size * 2];
+  memset(linked_hashes, 0x00, sizeof(linked_hashes));
+
+  h26x_nalu_list_print(nalu_list);
+
+  // Start with the first item in the NALU list and reset the list index.
+  item = nalu_list->first_item;
+  *list_idx = 0;
+  remove_used_in_gop_hash(self->nalu_list);  // Clear previous `used_in_gop_hash` flags.
```

Comment:
Why do you need to remove previously set flags? Isn't something wrong if there are pending items that have been tagged already?